### ISSUE TITLE
[cockpit/python]-bridge memory usage/leak (systemd pkg?)

### ISSUE THREAD

**Issue by jciesla** [cockpit-project/cockpit#19049]
Hello,

Hello, I currently have an issue for which I try to find the origin. It does not matter if it is for use with cockpit-bridge or python-bridge, when a user session remains open for a certain time, the consumption of the associated process constantly increases (cockpit-bridge or python), until a memory failure occurs (by example with oom killer).

I found that if I remove the systemd package from my system (delete the "/usr/share/cockpit/systemd"), I no longer haves the memory leak. All other packages working correctly without any issue, even with sessions maintained for long periods.

I try with each new cockpit version and I thought that the new python-bridge might be able to solve the situation but unfortunately the results are the same.

The last version I tried is 295.
I use a custom Linux distribution for industrial use, running kernel 5.19 (Linux 5.19.17-rt10 #1 SMP PREEMPT_DYNAMIC x86_64 x86_64 x86_64 GNU/Linux)

![image](https://github.com/cockpit-project/cockpit/assets/11711472/db1ef0af-813e-49d8-9772-c4193ad88572)

Maybe someone can give me an idea or help me to identify the issue. Thanks!

**Comment by jelly** [comment]
Some things which might be interesting for us to debug this issue:

* what systemd version do you use?
* what page do you keep open which reproduces the issue? Is it the overview page or different page?
* Are there any errors on your browsers console?

**Comment by jciesla** [comment]
Hello @jelly, 

This is the systemd version :

```
admin@hw-ovf21:~$ systemctl --version
systemd 250 (250.5+)
+PAM -AUDIT -SELINUX -APPARMOR +IMA -SMACK -SECCOMP -GCRYPT -GNUTLS -OPENSSL -ACL +BLKID -CURL -ELFUTILS -FIDO2 -IDN2 -IDN -IPTC +KMOD -LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE -BZIP2 -LZ4 +XZ -ZLIB -ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
admin@hw-ovf21:~$
```

I stripped down the most of cockpit folders to isolate the issue and restarted cockpit before to login :

```
root@hw-ovf21:/usr/share/cockpit# ls -l
total 32
drwxr-xr-x  2 root root 4096 Mar  9  2018 base1
drwxr-xr-x 14 root root 4096 Mar  9  2018 branding
drwxr-xr-x  2 root root 4096 Mar  9  2018 metrics
drwxr-xr-x  2 root root 4096 Mar  9  2018 motd
drwxr-xr-x  3 root root 4096 Mar  9  2018 shell
drwxr-xr-x  2 root root 4096 Mar  9  2018 ssh
drwxr-xr-x  3 root root 4096 Mar  9  2018 static
drwxr-xr-x  2 root root 4096 Mar  9  2018 systemd
root@hw-ovf21:/usr/share/cockpit#
```

When I log in, I use a user authorized on the system sudoers (admin), and administrative rights has been enabled. I only log into the overview page and do not click anywhere.

the [python]-bridge process leaking is the one associated to the admin user (the bridge for root user do not leak).
![image](https://github.com/cockpit-project/cockpit/assets/11711472/f1a8e10e-32ac-4bdc-97cf-6d4c6aabfeee)
![image](https://github.com/cockpit-project/cockpit/assets/11711472/c1937f73-b5db-4daf-aa09-f780b0d2a829)

On the browser, I see these errors.
![image](https://github.com/cockpit-project/cockpit/assets/11711472/1f9662ef-9953-4933-9643-1d4c8fdf1435)

**Comment by encbar5** [comment]
@jelly This does appear to still be an issue. I have an ARM system (aarch64) using Debian 12, and both the C bridge and python bridge appear to have leaks. Steps were to log in to a Cockpit session from a Firefox tab, make sure the session had administrative privileges, and then open a web terminal. No other interaction with the cockpit interface was necessary. The tab was then left open.

Using cockpit version `287.1` from bookworm, the `cockpit-bridge --privileged` process almost doubles the resident set in 50 minutes, from 7b to 14Mb.

```
Mon 15 Jul 19:26:30 CDT 2024
root        2578  1.8  0.2 310176  7836 ?        Sl   19:26   0:00 cockpit-bridge --privileged

Mon 15 Jul 20:17:24 CDT 2024
root        2578  0.1  0.3 311140 14108 ?        Sl   19:26   0:03 cockpit-bridge --privileged
```
Then testing the `320` version from bookworm-backports, it is using the python bridge, which started at higher memory, 28Mb. This did not change after 50 minutes, but then I clicked a few more buttons in the Cockpit UI, opened a web terminal, and left it for several more hours, which resulted in a resident set of 100Mb.

```
Mon 15 Jul 20:21:28 CDT 2024
root       10276  2.3  0.7  41648 28672 ?        S    20:20   0:00 /usr/bin/python3 /usr/bin/cockpit-bridge --privileged

Mon 15 Jul 21:14:08 CDT 2024
root       10276  0.1  0.7  41648 28672 ?        S    20:20   0:04 /usr/bin/python3 /usr/bin/cockpit-bridge --privileged

Tue 16 Jul 05:45:25 CDT 2024
root       10276  0.1  2.5 113988 100608 ?       S    Jul15   0:54 /usr/bin/python3 /usr/bin/cockpit-bridge --privileged
```

**Comment by neetabisht** [comment]
Hello Everyone,
I am also seeing similar type of issue.
My setup is simple where we have created one new cockpit UI page for our application and when cockpit page remain idle for longer time, we are facing more memory consumption issue. Due to this "system health is at bad state" error logs are coming on console.

```
[ability@dcsosrpmbuild systemd]$ uname -a
Linux dcsosrpmbuild 5.14.0-284.30.1.el9_2.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Aug 25 09:13:12 EDT 2023 x86_64 x86_64 x86_64 GNU/Linux
[ability@dcsosrpmbuild systemd]$ systemctl --version
systemd 252 (252-14.el9_2.3)
+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
[ability@dcsosrpmbuild systemd]$ microshift version
MicroShift Version: 4.14.32
Base OCP Version: 4.14.32
[ability@dcsosrpmbuild systemd]$
```

Could anyone please help. Thanks

**Comment by patsuhiko** [comment]
Hello,
same problem here. Using Debian 12 on a x64 machine and after 4 days with the GUI opened on a client the memory is almost full.
Plugins installed: networkmanager, storaged, podman
Page opened: tested with overview and memory
Versions:
cockpit 287.1-0+deb12u3
cockpit-bridge 287.1-0+deb12u3
cockpit-networkmanager 287.1-0+deb12u3
cockpit-packagekit 287.1-0+deb12u3
cockpit-pcp 287.1-0+deb12u3
cockpit-podman 67-1
cockpit-storaged 287.1-0+deb12u3
cockpit-system 287.1-0+deb12u3
cockpit-ws 287.1-0+deb12u3

Thank you.

**Comment by encbar5** [comment]
I had this experience also, that logging out of the admin session kills the privileged process and frees the memory.

**Comment by patsuhiko** [comment]
Yeah, this works everytime. My current fix is a timer that checks if the cockpit-bridge consumes more than 500mb memory and if so - the cockpit-bridge will be killed. Not the best way - but it works.
Maybe we can find the final solution where i can remove the timer :)

**Comment by tacaswell** [comment]
Possible duplicate of https://github.com/cockpit-project/cockpit/issues/13689

**Comment by tacaswell** [comment]
I saw a `cockpit-bridge` process get up over 1.6G of memory used.  I had a long-running browser tab open and closing the browser tab killed the process.

**Comment by mratsim** [comment]
Over 2.5G used here

<img width="1838" height="111" alt="Image" src="https://github.com/user-attachments/assets/ccd3b2cc-f9a6-4440-8bbf-5144f3a6525c" />

**Comment by martinpitt** [comment]
Indeed this is almost certainly a duplicate of #21468, i.e. should be fixed now.
